### PR TITLE
Plant part updates to fix RMI CI memory issues

### PR DESCRIPTION
**Pull request by katie-lamb**
Currently my RMI branch (`update-ci`) references this PUDL branch. This branch should get merged first before the RMI CI branch.

There are two changes that make the PPL more memory efficient:
- clears the `pudl_out._dfs` cache after all the plant parts list inputs are compiled and before the start of the plant parts list compilation
- converts columns in the plant parts list to strings or categoricals to be more memory efficient

Is this type of update worthy of including in the release notes??

### CODE REVIEW

File: src/pudl/metadata/fields.py [hunk 1]
```diff
@@ -2120,6 +2152,17 @@
 
 FIELD_METADATA_BY_RESOURCE: dict[str, dict[str, Any]] = {
     "sector_consolidated_eia": {"code": {"type": "integer"}},
+    "plant_parts_eia": {
```
Comment (katie-lamb) [inline src/pudl/metadata/fields.py:2155]: This isn't the nicest to make all the `category` columns part of the resource specific metadata but when I included categorical columns in the general `FIELD_METADATA` there were some errors from `string` columns becoming `category` so I decided it's safest to do all the categorical dtype conversions at the end on a resource specific level.

File: src/pudl/helpers.py [hunk 2]
```diff
@@ -1083,6 +1086,7 @@ def convert_cols_dtypes(
     Args:
         df: dataframe with columns that appear in the PUDL tables.
         data_source: the name of the datasource (eia, ferc1, etc.)
+        data_resource: the name of the resource (plant_parts_eia)
```
Comment (zaneselvans) [inline src/pudl/helpers.py:1089]: Why is this change necessary? Does it only apply to the PPE?

File: src/pudl/metadata/fields.py (hunk 1 above)
Comment (zaneselvans) [reply src/pudl/metadata/fields.py:2155]: It's probably more correct for a bunch of our coded columns to be categoricals in the database, but they're almost all strings right now, and I'm sure we rely on that typing in a ton of places, so this seems fine for now. I made an issue to remember this in, if you have thoughts to add: #1866

File: src/pudl/metadata/fields.py [hunk 3]
```diff
@@ -2143,8 +2186,10 @@
 
 def get_pudl_dtypes(
     group: str | None = None,
+    resource: str | None = None,
```
Comment (zaneselvans) [inline src/pudl/metadata/fields.py:2189]: Would this resource specific selection still be necessary if the PPE were listed as part of the EIA data source (which wouldn't seem crazy). Or I guess it's the need to override with the resource specific Categorical types.

Did you look at the `Resource.format_df()` method as an alternative by chance?

File: src/pudl/metadata/constants.py [hunk 4]
```diff
@@ -14,6 +14,7 @@
     "date": "datetime64[ns]",
     "datetime": "datetime64[ns]",
     "year": "datetime64[ns]",
+    "category": "category",
```
Comment (zaneselvans) [inline src/pudl/metadata/constants.py:17]: We might want to think about this a little bit in the context of how our resource definition & typing system works.

Right now we can set an `enum` constraint on a column, which results in a `CHECK VALUE IN () OR NULL` constraint being applied to the column when it's loaded into the database. How should that kind of constraint relate to the specification that a column is a `category` when it's loaded into a dataframe?

File: src/pudl/metadata/fields.py (hunk 3 above)
Comment (katie-lamb) [reply src/pudl/metadata/fields.py:2189]: I might have confused myself with the different metadata classes and functionality but walking through my thought process:
- `RESOURCE_METADATA` is the metadata for resources in the DB, so the PPE shouldn't be added as a resource here since it's just an output table. 
- I could use `Resource.format_df()` by defining a one off PPE resource and not add it to `RESOURCE_METADATA` but I think the only functionality I'd really be leveraging from `Resource.format_df()` is to set specific column types (in this case `category`), so I might as well use `df.astype()` to do that.
- In the future these fields will be `category` types as per #1866 and `apply_pudl_dtypes()` can be used without any group or resource overrides.
- For now, maybe these `category` type overrides should just happen in a helper function in the PPE module?
- Does it make sense to have any of the PPE fields be included in the `FIELD_METADATA` or should all the specific type conversion just happen in a PPE module helper function?

Comment (zaneselvans) [reply src/pudl/metadata/fields.py]: The metadata for the DB is defined by a `Package` which contains a list of `Resource` objects. So you could define`Resource` objects for the output tables, composed of fields that are defined in `fields.py` and just make sure that the output table `Resource`s aren't included in the DB definition.

**Comment by katie-lamb** [comment]
The latest iteration adds the plant parts list as a `Resource` and adds the attributes whose type I want to explicitly set as a `Field`. Then, `Resource.to_pandas_dtypes` retrieves the type of these attributes and they are set in `MakePlantParts`.

A few notes:
- Right now one can either:
  - Set an enum constraint on a column and set it to a python type. Later, `Field.to_pandas_dtype` will set this column to Pandas Categorical
  - Set a column type to a Pandas `category` with no enum constraint
- I thought there might be an instance where you want to set a column to `category` to save memory but don't actually care about constraining the values or updating the constraints when new values are introduced
- If this discrepancy is too weird, I could probably add the functionality to set value constraints on `category` types. Or it could be handled with #1866 
- Finally, I didn't add all the plant part list fields to `fields.py`, only the fields that needed their type explicitly set. Should I add the remaining 10 or so fields? I thought that I wouldn't require adding all the fields to the `Resource` because I've been continuously adding columns for experimenting with Panda. But maybe it's better practice to add all the fields in this output version of the plant part list.

File: src/pudl/metadata/fields.py [hunk 5]
```diff
@@ -2138,6 +2171,15 @@
             },
         }
     },
+    "plant_parts_eia": {
+        "energy_source_code_1": {
```
Comment (katie-lamb) [inline src/pudl/metadata/fields.py:2175]: The `energy_source_code_1` and `prime_movers_eia` resource override enum constraints could maybe just be moved to the general entry in `FIELD_METADATA`

Comment (zaneselvans) [reply src/pudl/metadata/fields.py:2175]: In the database tables these are both constrained by virtue of their FK relationships to the `energy_sources_eia` and `prime_movers_eia` coding tables, so adding the enums to the field definitions would be duplicative. But for the purpose of imposing those constraints on this free floating output table it seems like the ENUM constraint makes sense.

File: src/pudl/metadata/classes.py [hunk 6]
```diff
@@ -576,7 +576,7 @@ class Field(Base):
 
     name: SnakeCase
     type: Literal[  # noqa: A003
-        "string", "number", "integer", "boolean", "date", "datetime", "year"
+        "string", "number", "integer", "boolean", "date", "datetime", "year", "category"
```
Comment (zaneselvans) [inline src/pudl/metadata/classes.py:579]: `category` is not an allowed field type under the TableSchema definition, so this violates the referenced standard.

https://specs.frictionlessdata.io/table-schema/#types-and-formats

Under the TableSchema specification, this functionality is implemented with a type (e.g. `string`) plus an `enum` constraint that enumerates the values the field is allowed to take on, so we should probably figure out how to implement the desired functionality within that framework.

File: docs/release_notes.rst [hunk 7]
```diff
@@ -116,11 +116,19 @@ Plant Parts List Module Changes
...
   of the default list are needed from the EIA 860 generators table, these columns can be
   passed in with the ``gens_cols`` argument.  See :pr:`1550`
+* For greater memory efficiency, appropriate columns are now cast to string and
+  categorical types when the full plant parts list is created.
 
 Metadata
 ^^^^^^^^
 * Used the data source metadata class added in release 0.6.0 to dynamically generate
   the data source documentation (See :doc:`data_sources/index`). :pr:`1532`
+* Column attributes may now be the pandas ``Categorical`` data type. Using categorical
```
Comment (zaneselvans) [inline docs/release_notes.rst:126]: I think this will break the system as it's implemented right now.

Column types are independent of Pandas (and SQLite, and Parquet, etc) and are defined by the [TableSchema specificaton](https://specs.frictionlessdata.io/table-schema/#field-descriptors). We then define mappings from that generic TableSchema type to the specific types for the various outputs (Pandas, SQLite, pyarrow, etc.).

It seems like the desired functionality is an unenumerated categorical column? Where we're telling the system "Please treat this as a categorical value, but figure out what the categories are dynamically based on the contents of the column, without imposing any constraints on the values."

This seems like it should not be implemented in the column type (since there is no categorical type under the standard, just enum constrained values of a given type), but rather in the process of translating to the various output formats.  Somehow we need to communicate to the output that some columns should be treated as categoricals when they're put into a dataframe or parquet file, even if they aren't constrained.

**Review by zaneselvans — CHANGES_REQUESTED** [review]
I agree this dynamically applied categorical type would be useful, but let's see if we can find an easy way to implement it without violating the TableSchema spec.

File: src/pudl/metadata/fields.py (hunk 5 above)
Comment (katie-lamb) [reply src/pudl/metadata/fields.py:2175]: Ah right, makes sense.

File: src/pudl/metadata/classes.py (hunk 6 above)
Comment (katie-lamb) [reply src/pudl/metadata/classes.py:579]: Got it. I'll take out all the `category` types and instead add an `enum` constraint. These fields can be cast to categoricals later down the line, with `Field.to_pandas_dtype`

Comment (zaneselvans) [reply src/pudl/metadata/classes.py:579]: Yeah, but somehow we have to convey to `to_pandas_dtype()` that it should be treated as a categorical, not just a vanilla string.  I think that that's what happens if you've got an enum constrained string, so if you add the enum in the PPE overrides, it should work maybe?

**Comment by zaneselvans** [comment]
I think if you're defining a PPE `Resource` all of the fields that are part of the `Resource` have to be defined, don't they?  Or are you asking it to process dataframes that have columns (which you want to retain) that aren't part of the Resource?

I think the high level function that we eventually really want to be using here is `Resource.format_df()` -- rather than all of the lower level components. `format_df()` ensures that all and only the columns defined in the resource exist in the dataframe, and that they conform to the types as specified. Basically it's the final preparatory step for loading the dataframe into a database table.

If a column is useful and pretty stable -- assumed to exist in analyses and functions that we have in the repo -- then it should probably be in the schema with a type, etc.

**Comment by katie-lamb** [comment]
I've now taken out the `"type": "category"` option from the field metadata and added an `enum` constraint for every PPE column that should be a categorical. In `MakePlantParts` I use `Resource.format_df()` to cast these `enum` constrained columns to pandas categorical types. 

All columns from the PPE are now included in the field metadata.

I still need to figure out why some records have `operational_status_pudl` equal to `BU` and then take `BU` out of the valid `enum` values for this field.

**Review by zaneselvans — CHANGES_REQUESTED** [review]

File: src/pudl/metadata/fields.py [hunk 8]
```diff
@@ -535,6 +572,13 @@
         "type": "string",
         "description": "Account number, from FERC's Uniform System of Accounts for Electric Plant. Also includes higher level labeled categories.",
     },
+    "ferc_acct_name": {
+        "type": "string",
+        "description": "Name of FERC account, derived from technology description and prime mover code.",
+        "constraints": {
+            "enum": ["other", "hydro", "steam", "nuclear", "Other", "Steam"]
```
[the remark this answers is not in the document]
Comment (katie-lamb) [reply src/pudl/metadata/fields.py]: Mistake leftover from a buggy PPE version. Fixed this.

File: src/pudl/analysis/plant_parts_eia.py [hunk 9]
```diff
@@ -649,6 +650,9 @@ def execute(self, gens_mega):
 
         """
         # aggregate everything by each plant part
+        df_keys = list(self.pudl_out._dfs.keys())
+        for k in df_keys:
```
[the remark this answers is not in the document]
Comment (katie-lamb) [reply src/pudl/analysis/plant_parts_eia.py:659]: You can't modify the length of a dictionary while iterating through it which is why I made a separate variable for the keys. Maybe there's a better way to do this.

File: src/pudl/metadata/enums.py [hunk 10]
```diff
@@ -213,3 +213,35 @@
     "Unknown Code",  # Should be replaced with NA
 ]
 """Valid emissions measurement codes for the EPA CEMS hourly data."""
+
+TECH_DESCRIPTIONS: list[str] = [
```
[the remark this answers is not in the document]
Comment (katie-lamb) [reply src/pudl/metadata/enums.py]: I tried converting all the `list` enums to `set` which broke a bunch of stuff because some of the enums are used in for concatenation with a list. Just updated the enums I added (`TECH_DESCRIPTIONS` and `PLANT_PARTS`) to be sets.

File: src/pudl/metadata/fields.py [hunk 11]
```diff
@@ -53,6 +54,26 @@
...
+        "constraints": {
+            "enum": [
+                "plant",
+                "plant_unit",
+                "plant_prime_mover",
+                "plant_technology",
+                "plant_prime_fuel",
+                "plant_ferc_acct",
+                "plant_operating_year",
+                "plant_gen",
```
[the remark this answers is not in the document]
Comment (katie-lamb) [reply src/pudl/metadata/fields.py]: I got a circular import error when I tried to import `pudl.analysis.plant_parts_eia.PLANT_PARTS` into `pudl.metadata.fields.py` or `pudl.metadata.enums.py`. I added PLANT_PARTS as an enum to at least take out the duplication of this list in `fields.py`.

File: src/pudl/metadata/fields.py [hunk 12]
```diff
@@ -1470,19 +1535,47 @@
...
     "plant_name_pudl": {
         "type": "string",
         "description": "Plant name, chosen arbitrarily from the several possible plant names available in the plant matching process. Included for human readability only.",
     },
+    "plant_part": {
+        "type": "string",
+        "description": "The part of the plant a record corresponds to.",
+        "constraints": {
+            "enum": [
+                "plant",
```
[the remark this answers is not in the document]
Comment (katie-lamb) [reply src/pudl/metadata/fields.py]: See comment about circular import error.

File: src/pudl/metadata/fields.py [hunk 13]
```diff
@@ -1470,19 +1535,47 @@
...
+        "type": "string",
+        "description": "A semi-manually cleaned version of the freeform FERC 1 plant name.",
+    },
     "plant_name_eia": {"type": "string", "description": "Plant name."},
     "plant_name_ferc1": {
         "type": "string",
         "description": "Name of the plant, as reported to FERC. This is a freeform string, not guaranteed to be consistent across references to the same plant.",
     },
-    "plant_name_clean": {
+    "plant_name_new": {
```
[the remark this answers is not in the document]
Comment (katie-lamb) [reply src/pudl/metadata/fields.py]: Updated to `plant_name_ppe`.

File: src/pudl/metadata/resources/eia.py [hunk 14]
```diff
@@ -285,6 +285,54 @@
         "etl_group": "entity_eia",
         "field_namespace": "eia",
     },
+    "plant_parts_eia": {
```
[the remark this answers is not in the document]
Comment (katie-lamb) [reply src/pudl/metadata/resources/eia.py:288]: Filled out these resource fields.

File: src/pudl/metadata/fields.py [hunk 15]
```diff
@@ -1357,10 +1413,19 @@
             "pattern": r"^\d{5}$",
         },
     },
+    "ownership": {
+        "type": "string",
+        "description": "Whether each generator record is for one owner or represents a total of all ownerships.",
+        "constraints": {"enum": ["owned", "total"]},
+    },
```
[the remark this answers is not in the document]
Comment (katie-lamb) [reply src/pudl/metadata/fields.py]: Changed to `ownership_record_type`.

File: src/pudl/metadata/fields.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (katie-lamb) [reply src/pudl/metadata/fields.py]: I changed the `PLANT_PARTS` dictionary to an `OrderedDict` and removed `PLANT_PARTS_ORDERED`.

File: src/pudl/analysis/plant_parts_eia.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (zaneselvans) [reply src/pudl/analysis/plant_parts_eia.py:650]: Ah okay! That makes sense. I didn't know that.

File: docs/release_notes.rst [hunk 16]
```diff
@@ -132,16 +132,22 @@ Plant Parts List Module Changes
   :mod:`pudl.analysis.mcoe.DEFAULT_GENS_COLS`. If additional columns that are not part
   of the default list are needed from the EIA 860 generators table, these columns can be
   passed in with the ``gens_cols`` argument.  See :pr:`1550`
-* For greater memory efficiency, appropriate columns are now cast to string and
+* For memory efficiency, appropriate columns are now cast to string and
   categorical types when the full plant parts list is created. The resource and field
-  metadata is now included in the PUDL metadata.
+  metadata is now included in the PUDL metadata. See :pr:`1865`
+* For clarity and specificity, the ``plant_name_new`` column was renamed
+  ``plant_name_eia`` and the ``ownership`` column was renamed ``ownership_record_type``.
```
Comment (zaneselvans) [inline docs/release_notes.rst:139]: I think you mean `plant_name_ppe` here, not `plant_name_eia`

**Review by zaneselvans — APPROVED** [review]
Just one mistake I think in the release notes, referring to `plant_name_eia` rather than `plant_name_ppe`.

File: src/pudl/metadata/fields.py [hunk 17]
```diff
@@ -535,6 +562,11 @@
         "type": "string",
         "description": "Account number, from FERC's Uniform System of Accounts for Electric Plant. Also includes higher level labeled categories.",
     },
+    "ferc_acct_name": {
+        "type": "string",
+        "description": "Name of FERC account, derived from technology description and prime mover code.",
+        "constraints": {"enum": ["Hydraulic", "Nuclear", "Steam", "Other"]},
```
Comment (zaneselvans) [inline src/pudl/metadata/fields.py:568]: Is there a reason to go with the Title Case values here rather than standardizing to lower case like we do almost everywhere else? Are we not normalizing strings in this column? Where does it come from?

Comment (katie-lamb) [reply src/pudl/metadata/fields.py:568]: These values are read in in `pudl.helpers.get_eia_ferc_acct_map()`, which reads in `pudl.package_data.glue.ferc_acct_to_pm_tech_map.csv`. In this CSV all fields (`technology_description`, `prime_mover_code`, `ferc_acct_name`) use Title Case. This table is then merged with the MCOE table to create the mega generators table. TitleCase is maintained in the merge keys and merged on fields in `pudl.analysis.plant_parts_eia.get_gens_mega_table()`. Since all of the fields use TitleCase (and a few others in the plant parts list/MCOE table, I'm leaving this as is for now.
